src/main.rs

```diff
@@ -0,0 +1,464 @@
+    reread_partition_table(dest)?;
+    udev_settle()?;
+
+    // postprocess
+    if config.ignition.is_some() || config.firstboot_kargs.is_some() || config.platform.is_some() {
+        let mount = mount_boot(&config.device)?;
+        if config.ignition.is_some() {
+            write_ignition(&mount.mountpoint, config.ignition.as_ref().unwrap())?;
+        }
+        if config.firstboot_kargs.is_some() {
+            write_firstboot_kargs(&mount.mountpoint, config.firstboot_kargs.as_ref().unwrap())?;
+        }
+        if config.platform.is_some() {
+            write_platform(&mount.mountpoint, config.platform.as_ref().unwrap())?;
+        }
+    }
+
+    Ok(())
+}
+
+// Copy the image to disk and verify its signature.
+fn write_image(source: &mut ImageSource, dest: &mut File) -> Result<()> {
+    // wrap source for GPG verification
+    let mut verify_reader: Box<dyn Read> = {
+        if !source.signature.is_empty() {
+            Box::new(GpgReader::new(
+                &mut source.reader,
+                source.signature.as_slice(),
+            )?)
+        } else {
+            Box::new(&mut source.reader)
+        }
+    };
+
+    // wrap again for progress reporting
+    let have_length = source.length_hint.is_some();
+    let length_hint = source.length_hint.unwrap_or(0);
+    let mut position: u64 = 0;
+    let mut last_report = Instant::now();
+    let mut progress_reader = ProgressReader::new(&mut verify_reader, |progress: usize| {
+        position += progress as u64;
+        if last_report.elapsed() >= Duration::from_secs(1)
+            || (have_length && position == length_hint)
+        {
+            last_report = Instant::now();
+            if have_length {
+                print!(
+                    "> Read {}/{} ({}%)   \r",
+                    format_bytes(position),
+                    format_bytes(length_hint),
+                    100 * position / length_hint
+                );
+            } else {
+                print!("> Read {}   \r", format_bytes(position));
+            }
+            let _ = std::io::stdout().flush();
+        }
+    });
+
+    // Wrap in a BufReader so we can peek at the first few bytes for
+    // format sniffing.  Don't trust the content-type since the server
+    // may not be configured correctly, or the file might be local.
+    // Then wrap in a reader for decompression.
+    let mut buf_reader = BufReader::with_capacity(16384, &mut progress_reader);
+    let mut decompress_reader: Box<dyn Read> = {
+        let sniff = buf_reader.fill_buf().chain_err(|| "sniffing input")?;
+        if sniff.len() >= 2 && &sniff[0..2] == b"\x1f\x8b" {
+            Box::new(GzDecoder::new(buf_reader))
+        } else if sniff.len() >= 6 && &sniff[0..6] == b"\xfd7zXZ\x00" {
+            Box::new(XzDecoder::new(buf_reader))
+        } else {
+            Box::new(buf_reader)
+        }
+    };
+
+    // Cache the first MB of input and write zeroes instead.  This ensures
```

Comment:
The first MiB should cover all the disk headers because it's usually where the first partition starts on modern systems (x86 at least).  It's probably a larger buffer than necessary.

Credit to @imcleod for the idea!